File: mcp-code-intelligence-nodejs/src/orchestration/meta/execute-dynamic.ts
/**
 * ExecuteDynamicTool — execute tool with mapping check + fallback chain support.
 * KSA-66: Routes via bridge's execute_dynamic_tool for nested tools (mapping check first).
 * KSA-139: Post-execution hooks for KB cache population/invalidation.
 * Behavioral parity with Python execute_dynamic.py.
 */

import { OrchestrationEngine } from '../engine.js';
import { CacheSource } from '../cache/index.js';

/** Execute a tool by name — mapping → chain → single routing. */
export async function executeDynamic(engine: OrchestrationEngine, args: Record<string, any>): Promise<string> {
  const toolName = args.tool_name;
  if (!toolName) return JSON.stringify({ error: "Missing 'tool_name'" });
  const toolArgs = args.arguments ?? {};
  const agentName = args.agent_name ?? 'default';

  const mapping = engine.getToolMapping(toolName);
  if (mapping) return executeViaBridgeWithCache(engine, toolName, mapping, toolArgs, agentName);

  const chain = engine.getRegistry().getChain(toolName);
  if (chain) return executeChainWithCache(engine, chain, toolArgs, agentName);

  return executeSingleWithCache(engine, toolName, toolArgs, agentName);
}

/** Execute via nested server's execute_dynamic_tool (bridge pattern). */
async function executeViaBridge(
  engine: OrchestrationEngine, toolName: string, mapping: [string, string], args: Record<string, any>
): Promise<string> {
  const [serverName, originalName] = mapping;
  const bridgeArgs = { tool_name: originalName, arguments: args };
  try {
    const result = await engine.callChild(serverName, 'execute_dynamic_tool', bridgeArgs);
    engine.getRegistry().recordHit(toolName, 1);
    if (!isErrorResult(result)) engine.getRegistry().recordHit(toolName, 3);
    return result;
  } catch (e: any) {
    return JSON.stringify({ error: `Nested execute failed on ${serverName}: ${e.message}` });
  }
}

/** Execute through fallback chain — try each server in priority order. */
async function executeChain(engine: OrchestrationEngine, chain: any, args: Record<string, any>): Promise<string> {
  const errors: string[] = [];
  for (const entry of chain.entries) {
    const actualName = entry.toolName ?? chain.toolName;
    try {
      const result = await engine.callChild(entry.serverName, actualName, args);
      engine.getRegistry().recordHit(chain.toolName, 1);
      if (!isErrorResult(result)) engine.getRegistry().recordHit(chain.toolName, 3);
      return result;
    } catch (e: any) {
      errors.push(`${entry.serverName}: ${e.message}`);
    }
  }
  return JSON.stringify({ error: `Tool '${chain.toolName}' failed on all ${chain.entries.length} servers: [${errors.join(', ')}]` });
}

/** Execute on single server via normal routing. */
async function executeSingle(engine: OrchestrationEngine, toolName: string, args: Record<string, any>): Promise<string> {
  try {
    const result = await engine.route(toolName, args);
    engine.getRegistry().recordHit(toolName, 1);
    if (!isErrorResult(result)) engine.getRegistry().recordHit(toolName, 3);
    return result;
  } catch (e: any) {
    return JSON.stringify({ error: e.message });
  }
}

function isErrorResult(result: string): boolean {
  return result.trimStart().startsWith('{"error"') || result.slice(0, 100).includes('"error"');
}

/** Execute via bridge with KB cache hooks (KSA-139). */
async function executeViaBridgeWithCache(
  engine: OrchestrationEngine, toolName: string, mapping: [string, string],
  args: Record<string, any>, agentName: string
): Promise<string> {
  const result = await executeViaBridge(engine, toolName, mapping, args);
  if (!isErrorResult(result)) {
    fireCacheWrite(engine, toolName, mapping[0], agentName);
  } else {
    fireCacheInvalidate(engine, toolName, agentName, result);
  }
  return result;
}

/** Execute chain with KB cache hooks (KSA-139). */
async function executeChainWithCache(
  engine: OrchestrationEngine, chain: any, args: Record<string, any>, agentName: string
): Promise<string> {
  const result = await executeChain(engine, chain, args);
  const toolName = chain.toolName;
  if (!isErrorResult(result)) {
    const serverName = chain.entries?.[0]?.serverName ?? 'unknown';
    fireCacheWrite(engine, toolName, serverName, agentName);
  } else {
    fireCacheInvalidate(engine, toolName, agentName, result);
  }
  return result;
}

/** Execute single with KB cache hooks (KSA-139). */
async function executeSingleWithCache(
  engine: OrchestrationEngine, toolName: string, args: Record<string, any>, agentName: string
): Promise<string> {
  const result = await executeSingle(engine, toolName, args);
  if (!isErrorResult(result)) {
    const tool = engine.getRegistry().find(toolName);
    const serverName = tool?.source ?? 'unknown';
    fireCacheWrite(engine, toolName, serverName, agentName);
  } else {
    fireCacheInvalidate(engine, toolName, agentName, result);
  }
  return result;
}

/** Fire-and-forget: write to KB cache on success (KSA-139). */
function fireCacheWrite(engine: OrchestrationEngine, toolName: string, serverName: string, agentName: string): void {
  try {
    const writer = engine.getKbCacheWriter();
    const tool = engine.getRegistry().find(toolName);
    const description = tool?.definition?.description ?? '';
    const inputSchema = tool?.definition?.inputSchema ?? {};
    // Determine source — for now treat as DISCOVERED (first use populates cache)
    writer.onSuccess(toolName, serverName, description, inputSchema, agentName, CacheSource.DISCOVERED)
      .catch((e: any) => console.error(`[kb-cache] Write error: ${e.message}`));
  } catch {
    // Non-blocking — never fail execution due to cache
  }
}

/** Fire-and-forget: invalidate KB cache on failure (KSA-139). */
function fireCacheInvalidate(engine: OrchestrationEngine, toolName: string, agentName: string, errorResult: string): void {
  try {
    const invalidator = engine.getKbCacheInvalidator();
    invalidator.onFailure(toolName, agentName, errorResult)
      .catch((e: any) => console.error(`[kb-cache] Invalidate error: ${e.message}`));
  } catch {
    // Non-blocking
  }
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/orchestration/meta/execute_dynamic.py
"""ExecuteDynamicTool — execute tool with fallback chain support.

Behavioral parity with Kotlin ExecuteDynamicTool.kt.
execute_sync() is the entry point (SYNC), uses a thread internally for async calls.
"""

from __future__ import annotations

import asyncio
import concurrent.futures
import json
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ..engine import OrchestrationEngine

_EXECUTOR = concurrent.futures.ThreadPoolExecutor(max_workers=2, thread_name_prefix="exec-dyn")


def execute_sync(engine: OrchestrationEngine, args: dict) -> str:
    """Sync entry point — schedules on orchestration's event loop."""
    tool_name = args.get("tool_name")
    if not tool_name:
        return json.dumps({"error": "Missing 'tool_name'"})
    arguments = args.get("arguments", {})
    agent_name: str = args.get("agent_name", "default")

    loop = getattr(engine, '_orch_loop', None)
    if loop is None or loop.is_closed():
        return json.dumps({"error": "Orchestration event loop not available"})

    future = asyncio.run_coroutine_threadsafe(
        _execute(engine, tool_name, arguments, agent_name), loop
    )
    try:
        return future.result(timeout=60)
    except concurrent.futures.TimeoutError:
        return json.dumps({"error": f"Tool '{tool_name}' timed out (60s)"})
    except Exception as e:
        return json.dumps({"error": str(e)})


def _run_async(engine: OrchestrationEngine, tool_name: str, arguments: dict, agent_name: str = "default") -> str:
    """Run in a new event loop on the executor thread."""
    loop = asyncio.new_event_loop()
    try:
        return loop.run_until_complete(_execute(engine, tool_name, arguments, agent_name))
    finally:
        loop.close()


async def _execute(engine: OrchestrationEngine, tool_name: str, arguments: dict, agent_name: str = "default") -> str:
    """Async implementation — mapping check → delegate via nested execute_dynamic_tool."""
    mapping = engine.get_tool_mapping(tool_name)
    if mapping:
        server_name, original_name = mapping
        try:
            # Call the nested server's execute_dynamic_tool (not the tool directly)
            result = await engine.call_child(
                server_name, "execute_dynamic_tool",
                {"tool_name": original_name, "arguments": arguments},
                timeout_ms=60_000
            )
            engine.get_registry().record_hit(tool_name, 1)
            if not _is_error_result(result):
                engine.get_registry().record_hit(tool_name, 3)
                _fire_cache_write(engine, tool_name, server_name, agent_name)
            else:
                engine.get_registry().record_hit(tool_name, -10)
                _fire_cache_invalidate(engine, tool_name, agent_name)
            return result
        except Exception as e:
            engine.get_registry().record_hit(tool_name, -10)
            _fire_cache_invalidate(engine, tool_name, agent_name)
            return json.dumps({"error": f"Nested execute failed: {e}"})

    # No mapping — try chain or single routing
    chain = engine.get_registry().get_chain(tool_name)
    if chain:
        return await _execute_chain(engine, chain, arguments, agent_name)
    return await _execute_single(engine, tool_name, arguments, agent_name)


async def _execute_chain(engine: OrchestrationEngine, chain, arguments: dict, agent_name: str = "default") -> str:
    """Try each server in chain until one succeeds."""
    errors: list[str] = []
    for entry in chain.entries:
        actual_name = entry.tool_name or chain.tool_name
        try:
            result = await engine.call_child(entry.server_name, actual_name, arguments)
            engine.get_registry().record_hit(chain.tool_name, 1)
            if not _is_error_result(result):
                engine.get_registry().record_hit(chain.tool_name, 3)
                _fire_cache_write(engine, chain.tool_name, entry.server_name, agent_name)
            else:
                engine.get_registry().record_hit(chain.tool_name, -10)
                _fire_cache_invalidate(engine, chain.tool_name, agent_name)
            return result
        except Exception as e:
            errors.append(f"{entry.server_name}: {e}")
    tool_name = chain.tool_name
    engine.get_registry().record_hit(tool_name, -10)
    _fire_cache_invalidate(engine, tool_name, agent_name)
    error_msg = f"Tool '{tool_name}' failed on all {len(chain.entries)} servers: [{', '.join(errors)}]"
    return json.dumps({"error": error_msg})


async def _execute_single(engine: OrchestrationEngine, tool_name: str, arguments: dict, agent_name: str = "default") -> str:
    """Execute on single server via normal routing."""
    try:
        result = await engine.route(tool_name, arguments)
        engine.get_registry().record_hit(tool_name, 1)
        if not _is_error_result(result):
            engine.get_registry().record_hit(tool_name, 3)
            # Determine server_name from routing for cache write
            server_name = _resolve_server_name(engine, tool_name)
            _fire_cache_write(engine, tool_name, server_name, agent_name)
        else:
            engine.get_registry().record_hit(tool_name, -10)
            _fire_cache_invalidate(engine, tool_name, agent_name)
        return result
    except Exception as e:
        engine.get_registry().record_hit(tool_name, -10)
        _fire_cache_invalidate(engine, tool_name, agent_name)
        return json.dumps({"error": str(e)})


def _is_error_result(result: str) -> bool:
    """Check if result indicates an error response."""
    return result.lstrip().startswith('{"error"') or '"error"' in result[:100]


def _fire_cache_write(engine: OrchestrationEngine, tool_name: str, server_name: str | None, agent_name: str) -> None:
    """Fire KB cache write on successful execution (best-effort)."""
    try:
        writer_fn = getattr(engine, 'get_kb_cache_writer', None)
        if writer_fn is None:
            return
        writer = writer_fn()
        if writer is None:
            return
        tool = engine.get_registry().find(tool_name)
        writer.on_success(
            tool_name,
            server_name or "",
            tool.description if tool else "",
            tool.input_schema if tool else {},
            agent_name,
            "discovered",
        )
    except Exception:
        pass  # Best-effort, never fail execution for cache write


def _fire_cache_invalidate(engine: OrchestrationEngine, tool_name: str, agent_name: str) -> None:
    """Fire KB cache invalidation on error (best-effort)."""
    try:
        invalidator_fn = getattr(engine, 'get_kb_cache_invalidator', None)
        if invalidator_fn is None:
            return
        invalidator = invalidator_fn()
        if invalidator is None:
            return
        invalidator.on_error(tool_name, agent_name)
    except Exception:
        pass  # Best-effort, never fail execution for cache invalidation


def _resolve_server_name(engine: OrchestrationEngine, tool_name: str) -> str | None:
    """Resolve server name for a tool from registry (best-effort)."""
    try:
        tool = engine.get_registry().find(tool_name)
        if tool is not None:
            return getattr(tool, 'server_name', None)
    except Exception:
        pass
    return None
